prime-dotnet-webapi/Services/Rules/AutomaticAdjudicationRules.cs

```diff
@@ -233,4 +233,63 @@ public override Task<bool> ProcessRule(Enrollee enrollee)
             return Task.FromResult(true);
         }
     }
+
+    public class IsPotentialPaperEnrolleeReturnee : AutomaticAdjudicationRule
+    {
+        private readonly IEnrolleePaperSubmissionService _enrolleePaperSubmissionService;
+
+        public IsPotentialPaperEnrolleeReturnee(IEnrolleePaperSubmissionService enrolleePaperSubmissionService)
+        {
+            _enrolleePaperSubmissionService = enrolleePaperSubmissionService;
+        }
+        public override async Task<bool> ProcessRule(Enrollee enrollee)
+        {
+            var PaperEnrollees = await _enrolleePaperSubmissionService.GetPotentialPaperEnrolleeReturnees(enrollee.DateOfBirth);
+            var PotentialPaperEnrolleeGpid = await _enrolleePaperSubmissionService.GetLinkedGpid(enrollee.Id);
+            var paperEnrolleeMatchId = -1;
+
+            // Check if there's a match on a birthdate in paper enrollees, get all the ones that have a match
+            if (PaperEnrollees != null)
```

Comment:
It gets harder to reason about logic like this the deeper the `if` statements get nested. I find its a lot more readable to do early exits rather than nesting like this; compare
```c#
if (PaperEnrollees != null)
{
    // *** if yes and GPID is provided
    if (PotentialPaperEnrolleeGpid != null)
    {
        // *** *** Check if GPID match one of the paper enrolment
        foreach (var PaperEnrollee in PaperEnrollees)
        {
            if (
                    PaperEnrollee.DateOfBirth.Date == enrollee.DateOfBirth.Date
                    && PaperEnrollee.GPID == PotentialPaperEnrolleeGpid
                )
            {
                paperEnrolleeMatchId = PaperEnrollee.Id;
            }
        }

        if (paperEnrolleeMatchId == -1)
        {
            enrollee.AddReasonToCurrentStatus(StatusReasonType.PaperEnrolmentMismatch, $"User-Provided GPID: {PotentialPaperEnrolleeGpid}");
            await _enrolleePaperSubmissionService.LinkEnrolmentToPaperEnrolment(enrollee.Id, paperEnrolleeMatchId);
            return false;
        }
        // *** *** if match "auto enrol" and link to paper enrolment and confirm the linkage here
        if (!await _enrolleePaperSubmissionService.LinkEnrolmentToPaperEnrolment(enrollee.Id, paperEnrolleeMatchId, isConfirmed: true))
        {
            // If for any reason the database call fails, send to manual adjudication with reason "Unable to link enrollee to paper enrollee"
            enrollee.AddReasonToCurrentStatus(StatusReasonType.UnableToLinkToPpaperEnrolment, $"User-Provided GPID: {PotentialPaperEnrolleeGpid}");
            return false;
        }
        return true;
    }
    // *** if yes and GPID not provided - flag with "Possible match with paper enrolment"
    else
    {
        await _enrolleePaperSubmissionService.LinkEnrolmentToPaperEnrolment(enrollee.Id, paperEnrolleeMatchId);
        enrollee.AddReasonToCurrentStatus(StatusReasonType.PossiblePaperEnrolmentMatch);
        return false;
    }
}
```
to
```c#
if (PaperEnrollees == null)
{
    return true;

}

// *** if yes and GPID is provided
if (PotentialPaperEnrolleeGpid != null)
{
    // *** *** Check if GPID match one of the paper enrolment
    foreach (var PaperEnrollee in PaperEnrollees)
    {
        if (
                PaperEnrollee.DateOfBirth.Date == enrollee.DateOfBirth.Date
                && PaperEnrollee.GPID == PotentialPaperEnrolleeGpid
            )
        {
            paperEnrolleeMatchId = PaperEnrollee.Id;
        }
    }

    if (paperEnrolleeMatchId == -1)
    {
        enrollee.AddReasonToCurrentStatus(StatusReasonType.PaperEnrolmentMismatch, $"User-Provided GPID: {PotentialPaperEnrolleeGpid}");
        await _enrolleePaperSubmissionService.LinkEnrolmentToPaperEnrolment(enrollee.Id, paperEnrolleeMatchId);
        return false;
    }
    // *** *** if match "auto enrol" and link to paper enrolment and confirm the linkage here
    if (!await _enrolleePaperSubmissionService.LinkEnrolmentToPaperEnrolment(enrollee.Id, paperEnrolleeMatchId, isConfirmed: true))
    {
        // If for any reason the database call fails, send to manual adjudication with reason "Unable to link enrollee to paper enrollee"
        enrollee.AddReasonToCurrentStatus(StatusReasonType.UnableToLinkToPpaperEnrolment, $"User-Provided GPID: {PotentialPaperEnrolleeGpid}");
        return false;
    }
    return true;
}
// *** if yes and GPID not provided - flag with "Possible match with paper enrolment"
else
{
    await _enrolleePaperSubmissionService.LinkEnrolmentToPaperEnrolment(enrollee.Id, paperEnrolleeMatchId);
    enrollee.AddReasonToCurrentStatus(StatusReasonType.PossiblePaperEnrolmentMatch);
    return false;
}
```

Reply:
Also, `PaperEnrollees` will never be null, it will be an empty array if none exist (after calling `.ToListAsync()` in the service)

Reply:
Fixed